src/dachshund/search_problem.rs

```diff
@@ -0,0 +1,33 @@
+pub struct SearchProblem {
```

Comment:
This abstraction looks pretty good to me. It definitely cleans up a lot of the other code and makes it easier to understand what's being passed around. We should definitely merge this since it's a big step forward.

Longer term, does it make sense to break this down into the SearchProblem, (ie the things like global_thresh and local_thresh that define the search space and the admissible solutions) the SearchStrategy (things like the beam size and num_to_search that govern how we look for solutions).

Reply:
Yes, absolutely! I was definitely thinking we could expand this abstraction later down the road, to account for other kinds of search (e.g. A-star vs. Beam search vs. Reinforcement Learning etc. etc.). But, one step at a time!